### PR TITLE
Retry mechanism for the vulnerability detector scans and usage of the helpers to get packages and hotfixes in an atomic way

### PR DESCRIPTION
**Pull request by DProvinciani**
|Related issue|
|---|
| https://github.com/wazuh/wazuh/issues/8572 |

## Description

This pull request includes the next set of changes:

- The call to get the packages to be scanned in Vulnerability Detector now uses the atomic helper.
- The call to get the hotfixes to be used during the Vulnerability Detector scan now uses the atomic helper.
- The loop that iterates the agents that should be scanned now implements a five times retry mechanism for the agents for which the packages or hotfixes information could not be obtained.
- If Vulnerability Detector fails five times on the agent scan, it logs a warning message.

## Logs/Alerts example

A warning message is generated if after making five attempts to obtain the software to be scanned, this information cant be obtained.

```
(5507): The software of the agent '000' could not be obtained after 5 attempts. Skipping agent until the next scan.
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Package installation

- Memory tests for Linux
  - [x] Scan-build report
image

### CODE REVIEW

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c [hunk 1]
```diff
@@ -4802,86 +4798,22 @@ int wm_vuldet_collect_agent_software(scan_agent *agent, sqlite3 *db, scan_ctx_t*
...
-        } else {
-            goto end;
-        }
-    }
-
-    // Avoid checking the same packages again
-    snprintf(buffer, OS_SIZE_6144, vu_queries[VU_SYSC_UPDATE_SCAN], agent->agent_id, scan_id);
-    if (wm_vuldet_wdb_request(buffer, OS_SIZE_6144)) {
-        mtdebug1(WM_VULNDETECTOR_LOGTAG, VU_SOFTWARE_REQUEST_ERROR, scan_ctx->agent_id);
+    if (OS_INVALID == wm_vuldet_get_software(atoi(agent->agent_id), VU_PARTIAL_SCAN ? true : false, &package_list)) {
```
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:4801]: Maybe, we should make this comparison in the ternary operator 
`scan_ctx->scan_type == VU_PARTIAL_SCAN`

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c [hunk 2]
```diff
@@ -6783,71 +6716,42 @@ int wm_vuldet_insert_MSU(sqlite3 *db, vu_msu_entries *msu) {
...
-        goto end;
-    }
-
-    snprintf(request, OS_SIZE_128, vu_queries[VU_HOTFIXES_REQUEST], agent_id, last_id);
 
-    if (wm_vuldet_wdb_request(request, OS_MAXSTR)) {
+    if (OS_INVALID == wm_vuldet_get_hotfixes(atoi(agent_id), &hotfixes)) {
         goto end;
     }
+    else {
```
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:6725]: We could check, just like packages, the JSON pointer before using it 
`if (hotfixes)`

**Review by pereyra-m — COMMENTED** [review]
Just a quick review!

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c (hunk 1 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:4801]: Fixed

**Comment by palaciosjeremias** [comment]
Great that all these deprecated queries are cleaned!

```
    [SELECT_QUERY] = "SELECT %s FROM %s WHERE %s;",

```
								
[src/wazuh_modules/vulnerability_detector/wm_vuln_detector_db.h](https://github.com/wazuh/wazuh/blob/dd60ade0b69bf975e0c8c680dedac91cc5dbd667/src/wazuh_modules/vulnerability_detector/wm_vuln_detector_db.h#L176) (Line 176)

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c (hunk 2 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:6725]: Fixed

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c [hunk 3]
```diff
@@ -2342,86 +2341,90 @@ int wm_vuldet_check_agent_vulnerabilities(wm_vuldet_t *vuldet) {
...
+        pending_agents = false;
+        for (agents_it = vuldet->scan_agents; agents_it; agents_it = agents_it->next) {
+            scan_ctx_t scan_ctx = {0};
+            scan_ctx.agent_id = atoi(agents_it->agent_id);
+            scan_ctx.agent_name = agents_it->agent_name;
+            scan_ctx.agent_ip = agents_it->agent_ip;
+            mtinfo(WM_VULNDETECTOR_LOGTAG, VU_START_AG_AN, scan_ctx.agent_id);
+            time_t start = time(NULL);
+
+            if (agents_it->scanned || (!agents_it->scanned && agents_it->attempts == 0)) continue;
```
Comment (palaciosjeremias) [inline src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:2354]: I think we can simplify this by eliminating agents_it->scanned, and use a unique agents_it->pending_attempts that will be decreased after each failure scanand set to 0 after a successfull scan.

**Review by palaciosjeremias — COMMENTED** [review]
Great implementation. A lot of code improvements!
Just minor observations.

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c [hunk 4]
```diff
@@ -2330,8 +2328,9 @@ int wm_vuldet_check_agent_vulnerabilities(wm_vuldet_t *vuldet) {
     scan_agent *agents_it;
     sqlite3 *db;
     sqlite3_stmt *stmt = NULL;
+    bool pending_agents = false;
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:2331]: Fixed

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.h [hunk 5]
```diff
@@ -353,6 +354,8 @@ typedef struct scan_agent_flags {
 
 typedef struct scan_agent {
     char *agent_id;
+    bool scanned;
+    int attempts;
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector.h:358]: Fixed

File: src/error_messages/error_messages.h [hunk 6]
```diff
@@ -323,7 +323,7 @@
 #define VU_CHECK_DB_ERROR           "(5504): Database check failed."
 #define VU_OS_VERSION_ERROR         "(5505): Invalid OS version."
 #define VU_REFRESH_DB_ERROR         "(5506): Could not refresh the '%s' DB."
-#define VU_GET_SOFTWARE_ERROR       "(5507): The software of the agent '%.3d' could not be obtained."
+#define VU_GET_SOFTWARE_ERROR       "(5507): The software of the agent '%.3d' could not be obtained after five attempts. Skipping agent until the next scan."
```
[the remark this answers is not in the document]
Comment (DProvinciani) [reply src/error_messages/error_messages.h:326]: Fixed

File: src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c (hunk 3 above)
Comment (DProvinciani) [reply src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c:2354]: You are right! Fixed!

**Review by palaciosjeremias — APPROVED** [review]
LGTM!

**Pull request merged**